src/event-log.c

```diff
@@ -0,0 +1,398 @@
+
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *message = NULL;
+
+		/* print message */
+		if (g_strcmp0(ifield->key, "MESSAGE") == 0) {
+			message = ifield->value;
+			g_string_append_printf(gstring, "%s", message);
+			break;
+		}
+	}
+
+	/* skip other fields if non-verbose */
+	if (!verbose)
+		return g_string_free(gstring, FALSE);
+
+	/* collect verbose fields in order of appearance */
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *key = NULL;
+		const gchar *value = NULL;
+
+		if (g_strcmp0(ifield->key, "TRANSACTION_ID") == 0) {
+			key = "transaction ID";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BUNDLE_HASH") == 0) {
+			key = "bundle hash";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BOOT_ID") == 0) {
+			key = "boot ID";
+			value = ifield->value;
+		}
+
+		if (key) {
+			g_string_append_printf(gstring, "\n                      %s: %s", key, value);
+		}
+	}
+
+	return g_string_free(gstring, FALSE);
+}
+
+/**
+ * Rotates log files.
+ *
+ * If a maximum size is configured and the loggers current size + provided
+ * 'next_len' exceed this maximum size, the method will close the current
+ * output stream, and copy the file to a .1 copy.
+ *
+ * All previously existing rotation files will be rotated by +1.
+ *
+ * If max-files is set, rotation is limited to max-files files.
+ *
+ * @param logger Logger to rotate file for
+ * @param next_len Length of next data to add.
+ *        This is used to determine if the log file needs to be rotated.
+ */
+static void rotate_logfiles(REventLogger* logger, goffset next_len)
+{
+	g_autoptr(GFile) logfile = NULL;
+	g_autofree gchar *rotfile = NULL;
+	g_autoptr(GError) ierror = NULL;
+
+	g_return_if_fail(logger);
+
+	/* method is no-op if no maxsize is configured */
+	if (!logger->maxsize)
+		return;
+
+	/* return if size limit not reached yet */
+	if (logger->filesize + next_len <= logger->maxsize)
+		return;
+
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+	/* iterate through list of found existing files and delete the first
+	 * one(s) and move the others */
+	for (guint file_num = logger->maxfiles; file_num > 0; file_num--) {
+		g_autofree gchar *from_file = g_strdup_printf("%s.%d", logger->filename, file_num -1);
+		g_autofree gchar *to_file = g_strdup_printf("%s.%d", logger->filename, file_num);
+
+		/* skip non-existing files */
+		if (!g_file_test(from_file, G_FILE_TEST_EXISTS))
+			continue;
+
+		if (g_rename(from_file, to_file) == -1) {
+			int err = errno;
+			g_warning("Failed to rotate log file: %s", g_strerror(err));
+		}
+		continue;
+	}
+
+	/* rotate current log file as next .1 file */
+	rotfile = g_strdup_printf("%s.1", logger->filename);
+	if (g_rename(logger->filename, rotfile) == -1) {
+		int err = errno;
+		g_warning("Failed to rotate log file: %s", g_strerror(err));
+		return;
```

Comment:
In this case, we no longer have a log file open, right? So writing the next log event would fail?

Reply:
Two possible solutions I see here:

a) ignore that we failed to rename and continue with re-opening (the probably still existing) log file

b) abort and disable logger

the disadvantage of a) is that it will probably fail on each further iteration and exceeds `max-size`. The disadvantage of b) is that we do not log anymore despite we probably could.

Reply:
Done so.